### PR TITLE
Upgrade gql java 9

### PR DESCRIPTION
**Pull request by yarinvak**
Upgrading to graphql - java 9
+
Adding directives support:
Untill now annotations had no support for directives.
This pr adds a support for directives -> you can pass GraphQLDirectives via object method, as definitions.
You can wire directives to fields and types via ``@GraphQLDirectives`` annotation.
And you can supply a special wiring class that will wire your datafetchers and transform your types/fields according to specified directive.
An example will be supplied later, in the tests.
An inspiration for the wiring was taken from the graphql-java library, which added directive wiring support in the SDL in version 9

The PR is not fully ready yet but I've opened it so you can have your remarks
I still have to add tests and readme

### CODE REVIEW

File: src/test/java/graphql/annotations/processor/retrievers/fieldBuilders/DirectivesBuilderTest.java [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+        assertEquals(graphQLDirectives[0].getName(), "upperCase");
+        assertFalse(graphQLDirectives[0].getArguments().isEmpty());
+        assertEquals(graphQLDirectives[0].getArguments().get(0).getName(), "myArg");
+        assertEquals(graphQLDirectives[0].getArguments().get(0).getValue(), "This is a string");
+        assertEquals(graphQLDirectives[1].getName(), "lowerCase");
+        assertTrue(graphQLDirectives[1].getArguments().isEmpty());
+    }
+
+    @Test
+    public void testDecoratedFieldWithArguments_argumentsValuesLongerThanArgumentsNumber_exceptionIsThrown() throws Exception {
```
Comment (guy120494) [inline src/test/java/graphql/annotations/processor/retrievers/fieldBuilders/DirectivesBuilderTest.java]: https://stackoverflow.com/questions/2469911/how-do-i-assert-my-exception-message-with-junit-test-annotation
This is the way to test that exception is thrown

File: src/test/java/graphql/annotations/processor/retrievers/fieldBuilders/DirectivesBuilderTest.java [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+        try {
+            directivesBuilder.build();
+            throw new Exception();
+        } catch (GraphQLAnnotationsException e) {
+            assertEquals(e.getMessage(), "Directive 'upperCase' is supplied with more argument values than it supports");
+        }
+    }
+
+    @Test
+    public void testDecoratedFieldWithArguments_directiveArgumentIsNotAScalarType_exceptionIsThrown() throws Exception {
```
Comment (guy120494) [inline src/test/java/graphql/annotations/processor/retrievers/fieldBuilders/DirectivesBuilderTest.java]: Be consistent with the names. "exceptionIsThrown" or "throwAnException"

File: src/test/java/graphql/annotations/directives/DirectiveWiringMapRetrieverTest.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        try {
+            directiveWiringMapRetriever.getDirectiveWiringMap(this.getClass().getField("field"), container);
+            throw new Exception();
+        } catch (GraphQLAnnotationsException e) {
+            assertEquals(e.getMessage(), "No directive named upperCase is found in the directive registry");
+        }
+    }
+
+    @Test
+    public void getDirectiveWiringMap_wiringClassIsPrivate_throwAnException() throws Exception {
```
Comment (guy120494) [inline src/test/java/graphql/annotations/directives/DirectiveWiringMapRetrieverTest.java]: Which exception is thrown? specify it in the test's name

File: src/test/java/graphql/annotations/directives/DirectiveWirerTest.java [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+        // Act
+        directiveWirer.wire(directiveContainer, map);
+
+        // Assert
+
+        verify(upperWiring).onInputObjectField(upperCaseEnv);
+    }
+
+    @Test(expectedExceptions = InvalidDirectiveLocationException.class)
+    public void wireGraphQLInputObjectField_invalidLocations_exception() throws Exception {
```
Comment (guy120494) [inline src/test/java/graphql/annotations/directives/DirectiveWirerTest.java]: Be consistent with names of tests that throw exception

File: src/main/java/graphql/annotations/processor/retrievers/fieldBuilders/DirectivesBuilder.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+public class DirectivesBuilder implements Builder<GraphQLDirective[]> {
+    private AnnotatedElement object;
+    private ProcessingElementsContainer container;
+
+    public DirectivesBuilder(AnnotatedElement object, ProcessingElementsContainer container) {
+        this.object = object;
+        this.container = container;
+    }
+
+    private GraphQLDirective transformArgs(GraphQLDirective graphQLDirective, String[] argumentValues) {
```
Comment (guy120494) [inline src/main/java/graphql/annotations/processor/retrievers/fieldBuilders/DirectivesBuilder.java:40]: Split this method

File: src/main/java/graphql/annotations/directives/DirectiveWirer.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+                element = wiring
+                        .onInputObjectType(new AnnotationsWiringEnvironmentImpl<>((GraphQLInputObjectType) element, element.getDirective(graphQLDirective.getName())));
+            } else if (element instanceof GraphQLInputObjectField) {
+                assertLocation(graphQLDirective, element, Introspection.DirectiveLocation.INPUT_FIELD_DEFINITION);
+                element = wiring
+                        .onInputObjectField(new AnnotationsWiringEnvironmentImpl<>((GraphQLInputObjectField) element, element.getDirective(graphQLDirective.getName())));
+            }
+        }
+        return element;
+
```
Comment (guy120494) [inline src/main/java/graphql/annotations/directives/DirectiveWirer.java]: Delete blank line

File: src/main/java/graphql/annotations/directives/DirectiveWirer.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+import java.util.Map;
+
+public class DirectiveWirer {
+    public GraphQLDirectiveContainer wire(GraphQLDirectiveContainer element, HashMap<GraphQLDirective, AnnotationsDirectiveWiring> directiveWiringMap) {
+        for (Map.Entry<GraphQLDirective, AnnotationsDirectiveWiring> entry : directiveWiringMap.entrySet()) {
+            System.out.println(entry.getKey());
+            GraphQLDirective graphQLDirective = entry.getKey();
+            AnnotationsDirectiveWiring wiring = entry.getValue();
+
+            if (element instanceof GraphQLFieldDefinition) {
```
Comment (guy120494) [inline src/main/java/graphql/annotations/directives/DirectiveWirer.java]: Change to Map<Class, Function> which the key is "GraphQLFieldDefinition,.class" and so on, and the value is the function inside the if statement

**Review by guy120494 — CHANGES_REQUESTED** [review]

File: src/main/java/graphql/annotations/directives/Directive.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target({ElementType.METHOD, ElementType.FIELD, ElementType.PARAMETER, ElementType.TYPE})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface Directive {
+    //    Class<? extends BasicDirectiveInfo> info();
```
Comment (osher-sade) [inline src/main/java/graphql/annotations/directives/Directive.java]: Remove redundant comment

**Pull request merged**